Hold the Scroll to the Text Fragment until a user scroll happens.

megangardner opened the pull request:
#### 4dc689b8cf6e04d32ec66f41d8c3d9615a7ec4a0
<pre>
Hold the Scroll to the Text Fragment until a user scroll happens.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245207">https://bugs.webkit.org/show_bug.cgi?id=245207</a>
rdar://99668200

Reviewed by Tim Horton.

For each layout that comes in until the user scrolls we need to reset
the scroll to keep the fragment in the middle of the page.
Otherwise, more content can come in the scroll the fragment off the
page, which can be very confusing to the user.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::reset):
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::maintainScrollPositionAtRange):
(WebCore::FrameView::textFragmentIndicatorTimerFired):
(WebCore::FrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::FrameView::scrollToRange):
(WebCore::FrameView::performPostLayoutTasks):
* Source/WebCore/page/FrameView.h:

Canonical link: <a href="https://commits.webkit.org/254507@main">https://commits.webkit.org/254507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7efddecad0c2c3ffe68d242abcbaf4c57c1b9a31

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33849 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/20079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98599 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/154922 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32354 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/27891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81652 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93054 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94937 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/25693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/76245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25636 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80563 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/20079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68612 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30128 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/20079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29855 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/20079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3163 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33302 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/76245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32007 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/20079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2321,6 +2322,19 @@ void FrameView::maintainScrollPositionAtAnchor(ContainerNode* anchorNode)
 
+void FrameView::maintainScrollPositionAtRange(SimpleRange* range)
+{
+    UNUSED_PARAM(range);
```
Get rid of this, since you clearly use it.

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2509,9 +2522,12 @@ void FrameView::textFragmentIndicatorTimerFired()
     m_pendingTextFragmentIndicatorRange.reset();
     m_pendingTextFragmentIndicatorText = String();
     m_delayedTextFragmentIndicatorTimer.stop();
+    
```
?

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -3531,6 +3547,30 @@ void FrameView::scrollToAnchor()
     cancelScheduledScrolls();
 }
 
+void FrameView::scrollToRange()
```
This is a super vague name. Can we call it "scrollToTextFragment"?

hortont424 replied:
(or scrollToTextFragmentRange, or something)

hortont424 commented on Source/WebCore/page/FrameView.h:
```diff
@@ -472,6 +472,7 @@ class FrameView final : public ScrollView {
 
     bool scrollToFragment(const URL&);
     void maintainScrollPositionAtAnchor(ContainerNode*);
+    void maintainScrollPositionAtRange(SimpleRange*);
```
ditto here, not just "range" but more importantly "text fragment"

hortont424 commented on Source/WebCore/page/FrameView.h:
```diff
@@ -911,6 +913,7 @@ class FrameView final : public ScrollView {
     RefPtr<Node> m_nodeToDraw;
     std::optional<SimpleRange> m_pendingTextFragmentIndicatorRange;
     String m_pendingTextFragmentIndicatorText;
+    bool m_doNotResetFragmentTimer { false };
```
I think this isn't just about the timer. It's "should incoming scrolling cancel maintaining scroll-to-text-fragment position". Try to make a better name. It's hard to compress down.

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2321,6 +2322,18 @@ void FrameView::maintainScrollPositionAtAnchor(ContainerNode* anchorNode)
         scrollToAnchor();
 }
 
+void FrameView::maintainScrollPositionAtScrollToTextFragmentRange(SimpleRange* range)
```
Why doesn't this take a reference?

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2321,6 +2322,18 @@ void FrameView::maintainScrollPositionAtAnchor(ContainerNode* anchorNode)
+{
+    LOG(Scrolling, "FrameView::maintainScrollPositionAtScrollToTextFragmentRange at %p", range);
+
+    m_pendingTextFragmentIndicatorRange = *range;
```
Since clearly it must be nonnull

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2509,6 +2521,8 @@ void FrameView::textFragmentIndicatorTimerFired()
 
 void FrameView::cancelScheduledTextFragmentIndicatorTimer()
 {
+    if (m_skipScrollResetofScrollToTextFragementRange)
```
Your camel casing fell apart in the middle

hortont424 replied:
Also your spelling of fragment

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -3531,6 +3545,30 @@ void FrameView::scrollToAnchor()
+    ASSERT(frame().document());
+    Ref document = *frame().document();
+
+    SetForScope doNotResetFragmentTimer(m_skipScrollResetofScrollToTextFragementRange, true);
```
The local name should be fixed up too

megangardner committed:
Hold the Scroll to the Text Fragment until a user scroll happens.
https://bugs.webkit.org/show_bug.cgi?id=245207
rdar://99668200

Reviewed by Tim Horton.

For each layout that comes in until the user scrolls we need to reset
the scroll to keep the fragment in the middle of the page.
Otherwise, more content can come in the scroll the fragment off the
page, which can be very confusing to the user.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::reset):
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::maintainScrollPositionAtRange):
(WebCore::FrameView::textFragmentIndicatorTimerFired):
(WebCore::FrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::FrameView::scrollToRange):
(WebCore::FrameView::performPostLayoutTasks):
* Source/WebCore/page/FrameView.h:

Canonical link: https://commits.webkit.org/254507@main
